Firestore/core/src/firebase/firestore/util/status.h

```diff
@@ -128,9 +144,49 @@ class ABSL_MUST_USE_RESULT Status {
     // NSError* to Status and back to NSError* losslessly.
     std::unique_ptr<PlatformError> platform_error;
   };
-  // OK status has a `nullptr` state_.  Otherwise, `state_` points to
-  // a `State` structure containing the error code and message(s)
-  std::unique_ptr<State> state_;
+
+  // A `unique_ptr` with a custom deleter. If the pointer's value has been set
+  // to a special value (0x1) to indicate it is moved, the custom deleter will
+  // stop `unique_ptr` trying to destrcut from an invalid address.
+  using StatePtr = std::unique_ptr<State, std::function<void(State*)>>;
+
+  static void CustomStateDelete(State* s) {
+    if (reinterpret_cast<uintptr_t>(s) != 0x1) {
+      delete s;
+    }
+  }
+
+  static StatePtr MakeStatePtrWithNull() {
+    return StatePtr(nullptr, &Status::CustomStateDelete);
+  }
+
+  static StatePtr MakeStatePtrWithMovedIndicator() {
+    return StatePtr(reinterpret_cast<State*>(0x1), &Status::CustomStateDelete);
+  }
+
+  static bool IsStatePtrMoved(const StatePtr& ptr) {
+    return reinterpret_cast<uintptr_t>(ptr.get()) == 0x1;
+  }
+
+  static StatePtr MakeStatePtr(const State& state) {
+    return StatePtr(new State(state), &Status::CustomStateDelete);
+  }
+
+  template <typename... Args>
+  static StatePtr MakeStatePtr(Args... args) {
+    return StatePtr(new State(args...), &Status::CustomStateDelete);
+  }
+
+  // OK status has a `nullptr` state_, and `IsMoved() == false`.  Otherwise,
+  // either `state_` points to a `State` structure containing the error code
+  // and message(s), or this instance has been moved.
+  StatePtr state_;
+
+  // Whether this instance has been moved. Without this `ok()` will be true
+  // for moved instances.
+  bool IsMoved() const;
```

Comment:
After implementing the rest of the feedback, I think this function is no longer used.